staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation.go

```diff
@@ -92,34 +93,49 @@ func validator(s *schema.Structural, isResourceRoot bool) *Validator {
 }
 
 // Validate validates all x-kubernetes-validations rules in Validator against obj and returns any errors.
```

Comment:
the contract around `costBudget` input and `remainingBudget` output needs to be super clear, especially about whose responsibility it is to pay attention to a negative number and error. The contact seems to be that if evaluating all rules would exceed costBudget (or if evaluations cannot be determined to have stayed in costBudget? see comments about the evaluation error and missing ActualCost cases), subsequent evaluations will be skipped, the list of `errs` returned will not be empty, and a negative number will be returned.

If that's the contract, then most callers can ignore the returned `remainingBudget` value unless they were going to make another validate call and needed to propagate the remainder on, right?

Reply:
That makes sense. I have added comments here.